x/ccv/provider/keeper/misbehaviour.go

```diff
@@ -0,0 +1,165 @@
+	// construct the trusted and conflicted light blocks
+	header1, err := headerToLightBlock(*misbehaviour.Header1)
+	if err != nil {
+		return nil, err
+	}
+	header2, err := headerToLightBlock(*misbehaviour.Header2)
+	if err != nil {
+		return nil, err
+	}
+
+	var validators []*tmtypes.Validator
+
+	// compare the signatures of the headers
+	// and return the intersection of validators who signed both
+
+	// create a map with the validators' address that signed header1
+	header1Signers := map[string]struct{}{}
+	for _, sign := range header1.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		header1Signers[sign.ValidatorAddress.String()] = struct{}{}
+	}
+
+	// iterate over the header2 signers
+	// and check if they s signed header1
+	for _, sign := range header2.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		if _, ok := header1Signers[sign.ValidatorAddress.String()]; ok {
+			_, val := header1.ValidatorSet.GetByAddress(sign.ValidatorAddress)
+			validators = append(validators, val)
+		}
+	}
+
+	sort.Sort(tmtypes.ValidatorsByVotingPower(validators))
+	return validators, nil
+}
+
+// headerToLightBlock returns a CometBFT light block from the given IBC header
+func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
+	sh, err := tmtypes.SignedHeaderFromProto(h.SignedHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	vs, err := tmtypes.ValidatorSetFromProto(h.ValidatorSet)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tmtypes.LightBlock{
+		SignedHeader: sh,
+		ValidatorSet: vs,
+	}, nil
+}
+
+// CheckMisbehaviour checks that headers in the given misbehaviour forms
+// a valid light client attack and that the corresponding light client isn't expired
+func (k Keeper) CheckMisbehaviour(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) error {
+	if err := misbehaviour.ValidateBasic(); err != nil {
+		return err
+	}
+
+	clientState, found := k.clientKeeper.GetClientState(ctx, misbehaviour.GetClientID())
+	if !found {
+		return sdkerrors.Wrapf(ibcclienttypes.ErrClientNotFound, "cannot check misbehaviour for client with ID %s", misbehaviour.GetClientID())
+	}
+
+	clientStore := k.clientKeeper.ClientStore(ctx, misbehaviour.GetClientID())
+
+	// Check that the headers are at the same height to ensure that
+	// the misbehaviour is for a light client attack and not a time violation,
+	// see ibc-go/modules/light-clients/07-tendermint/types/misbehaviour_handle.go#L54
+	if !misbehaviour.Header1.GetHeight().EQ(misbehaviour.Header2.GetHeight()) {
+		return sdkerrors.Wrap(ibcclienttypes.ErrInvalidMisbehaviour, "headers are not at same height")
+	}
+
+	// CheckMisbehaviourAndUpdateState verifies the misbehaviour against the consensus states
+	// but does NOT update the light client status.
+	// Note CheckMisbehaviourAndUpdateState returns an error if the light client is expired
+	_, err := clientState.CheckMisbehaviourAndUpdateState(ctx, k.cdc, clientStore, &misbehaviour)
```

Comment:
Will this eventually be based on ibc-go `v7.1.0`? This method was removed there and will look something like this:
```suggestion
	_, err := clientState.VerifyClientMessage(ctx, k.cdc, clientStore, &misbehaviour)
```
Also some checks have changed and it would be good if we can rebase.

Reply:
The current version is targeting ICS v2.x, which uses SDK 0.45 and IBC 4.2